platform/kube/controller.go

```diff
@@ -352,33 +361,30 @@ func (c *Controller) HostInstances(addrs map[string]bool) []*model.ServiceInstan
 	return out
 }
 
-const (
-	// the URI scheme used to encode a Kubernetes service account
-	uriScheme = "spiffe"
-)
-
 // GetIstioServiceAccounts returns the Istio service accounts running a serivce
 // hostname. Each service account is encoded according to the SPIFFE VSID spec.
 // For example, a service account named "bar" in namespace "foo" is encoded as
 // "spiffe://cluster.local/ns/foo/sa/bar".
 func (c *Controller) GetIstioServiceAccounts(hostname string, ports []string) []string {
 	saSet := make(map[string]bool)
+
+	// Get the service accounts running service within Kubernetes. This is reflected by the pods that
+	// the service is deployed on, and the service accounts of the pods.
 	for _, si := range c.Instances(hostname, ports, model.TagsList{}) {
-		key, exists := c.pods.keys[si.Endpoint.Address]
-		if !exists {
-			continue
-		}
-		item, exists, err := c.pods.informer.GetStore().GetByKey(key)
-		if !exists {
-			continue
-		}
-		if err != nil {
-			glog.V(2).Infof("Error retrieving pod by key: %v", err)
-			continue
+		if si.ServiceAccount != "" {
+			saSet[si.ServiceAccount] = true
 		}
+	}
 
-		pod, _ := item.(*v1.Pod)
-		sa := generateServiceAccountID(pod.Spec.ServiceAccountName, pod.GetNamespace(), c.domainSuffix)
+	// Get the service accounts running the service, if it is deployed on VMs. This is retrieved
+	// from the service annotation explicitly set by the operators.
+	svc, exists := c.GetService(hostname)
```

Comment:
This GetSerivce() returns a service specified by hostname. It can be VM related or K8s. At line 386, we are getting the service accounts from the service annotation, which is for the VMs. The above for loop gets the service accounts of the K8s pods that are running the service.